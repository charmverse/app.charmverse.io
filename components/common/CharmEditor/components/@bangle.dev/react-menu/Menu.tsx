import styled from '@emotion/styled';
import Paper from '@mui/material/Paper';
import React from 'react';

const StyledMenu = styled(Paper)`
  display: flex;
  padding: ${({ theme }) => theme.spacing(0.75, 0.5)};
`;

export function Menu ({
  className = '',
  children
}: {
  className?: string;
  children: React.ReactNode;
}) {
  return (
    <StyledMenu className={className} elevation={8}>
      {children}
    </StyledMenu>
  );
}
