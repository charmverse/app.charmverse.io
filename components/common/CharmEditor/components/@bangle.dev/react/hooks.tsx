import type { EditorView } from '@bangle.dev/pm';
import { Plugin, PluginKey } from '@bangle.dev/pm';
import { rafSchedule } from '@bangle.dev/utils';
import { useContext, useEffect, useState } from 'react';

import type { BangleEditorStateProps } from 'components/common/CharmEditor/components/@bangle.dev/core/bangle-editor-state';

import { BangleEditorState } from '../core/bangle-editor-state';

import { EditorViewContext } from './ReactEditor';

export function useEditorState(props: BangleEditorStateProps) {
  if (props.plugins && typeof props.plugins !== 'function') {
    throw new Error('plugins error: plugins must be a function');
  }

  const [state] = useState(
    () =>
      // Instantiate the editorState once and keep using that instance
      // on subsequent renders.
      // Passing a callback in useState lazy calls the
      // functions on the first render and never again.
      new BangleEditorState(props)
  );

  return state;
}

export function usePluginState(pluginKey: PluginKey, throttle = false) {
  const view = useEditorViewContext();
  const [state, setState] = useState(pluginKey.getState(view.state));

  useEffect(() => {
    let _setState = setState;
    if (throttle) {
      _setState = rafSchedule(setState);
    }
    const plugin = watcherPlugin(pluginKey, _setState);
    (view as any)._updatePluginWatcher(plugin);
    return () => {
      if (throttle) {
        (_setState as ReturnType<typeof rafSchedule>).cancel();
      }
      (view as any)._updatePluginWatcher(plugin, true);
    };
  }, [view, pluginKey, throttle]);

  return state;
}

export function useEditorViewContext(): EditorView {
  const context = useContext(EditorViewContext);
  if (!context) {
    throw new Error('useEditorViewContext must be used within a EditorViewContext');
  }
  return context;
}

function watcherPlugin(pluginKey: PluginKey, setState: (state: any) => void) {
  return new Plugin({
    key: new PluginKey(`withPluginState_${(pluginKey as any).key}`),
    view() {
      return {
        update(view, prevState) {
          const { state } = view;
          if (prevState === state) {
            return;
          }
          const newPluginState = pluginKey.getState(state);

          if (newPluginState !== pluginKey.getState(prevState)) {
            setState(newPluginState);
          }
        }
      };
    }
  });
}
