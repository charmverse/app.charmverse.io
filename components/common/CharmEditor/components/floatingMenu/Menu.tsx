import styled from '@emotion/styled';
import Paper from '@mui/material/Paper';
import React from 'react';

import type { SubMenu } from './floating-menu';

const StyledMenu = styled(Paper, { shouldForwardProp: (prop: string) => prop !== 'noScroll' })<{
  type?: SubMenu;
  noScroll?: boolean;
  inline?: boolean;
}>`
  display: flex;
  border-radius: 4px;
  ${({ theme, type }) =>
    type && ['defaultMenu', 'inlineCommentSubMenu'].includes(type) && theme.breakpoints.down('sm')} {
    ${({ inline, theme }) => !inline && `width: calc(100vw - ${theme.spacing(1)})`};
    margin: ${({ theme }) => theme.spacing(2, 0.5)};
    ${({ noScroll }) => (noScroll ? '' : 'overflow-x: auto')};
  }
`;

export function Menu({
  children,
  type,
  noScroll,
  inline,
  'data-test': dataTest
}: {
  children: React.ReactNode;
  type?: SubMenu;
  noScroll?: boolean;
  inline?: boolean;
  'data-test'?: string;
}) {
  return (
    <StyledMenu elevation={8} type={type} noScroll={noScroll} inline={inline} data-test={dataTest}>
      {children}
    </StyledMenu>
  );
}
