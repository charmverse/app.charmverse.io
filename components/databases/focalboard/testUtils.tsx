// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IntlProvider} from 'react-intl'
import React from 'react'
import {DndProvider} from 'react-dnd'
import {HTML5Backend} from 'react-dnd-html5-backend'
import configureStore, {MockStoreEnhanced} from 'redux-mock-store'
import {Middleware} from 'redux'

import {Block} from './blocks/block'

export const wrapIntl = (children?: React.ReactNode): JSX.Element => <IntlProvider locale='en'>{children}</IntlProvider>
export const wrapDNDIntl = (children?: React.ReactNode): JSX.Element => {
    return (
        <DndProvider backend={HTML5Backend}>
            {wrapIntl(children)}
        </DndProvider>
    )
}

export function mockDOM(): void {
    window.focus = jest.fn()
    document.createRange = () => {
        const range = new Range()
        range.getBoundingClientRect = jest.fn()
        range.getClientRects = () => {
            return {
                item: () => null,
                length: 0,
                [Symbol.iterator]: jest.fn(),
            }
        }
        return range
    }
}
export function mockMatchMedia(result: {matches: boolean}): void {
    // We check if system preference is dark or light theme.
    // This is required to provide it's definition since
    // window.matchMedia doesn't exist in Jest.
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation(() => {
            return result

            // return ({
            //     matches: true,
            // })
        }),
    })
}

export function mockStateStore(middleware:Middleware[], state:unknown): MockStoreEnhanced<unknown, unknown> {
    const mockStore = configureStore(middleware)
    return mockStore(state)
}

export type BlocksById<BlockType> = {[key: string]: BlockType}

export function blocksById<BlockType extends Block>(blocks: Array<BlockType>): BlocksById<BlockType> {
    return blocks.reduce((res, block) => {
        res[block.id] = block
        return res
    }, {} as BlocksById<BlockType>)
}
