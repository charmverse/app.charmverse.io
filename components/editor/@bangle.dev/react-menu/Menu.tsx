import styled from "@emotion/styled";
import React from 'react';
import { sidebarBackgroundColor, sidebarBackgroundColorDarkMode } from "theme/colors";

const StyledMenu = styled.div`
  background-color: ${({ theme }) => theme.palette.mode === "dark" ? sidebarBackgroundColorDarkMode : sidebarBackgroundColor};
  box-shadow: 0 10px 15px -3px rgb(0 0 0 / 0.1), 0 4px 6px -4px rgb(0 0 0 / 0.1);
  border-radius: ${({ theme }) => theme.spacing(0.5)};
  display: flex;
  padding: ${({ theme }) => theme.spacing(0.75, 0.5)};
`

export function Menu({
  className = '',
  children
}: {
  className?: string;
  children: React.ReactNode;
}) {
  return (
    <StyledMenu className={className}>
      {children}
    </StyledMenu>
  );
}
