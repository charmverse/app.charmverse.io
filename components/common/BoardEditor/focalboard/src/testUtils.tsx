import type { PageMeta } from '@charmverse/core/pages';
import React from 'react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { IntlProvider } from 'react-intl';
import type { Middleware } from 'redux';
import type { MockStoreEnhanced } from 'redux-mock-store';
import configureStore from 'redux-mock-store';
import { v4 } from 'uuid';

import { PagesContext } from 'hooks/usePages';
import type { Block } from 'lib/focalboard/block';
import { pageStubToCreate } from 'testing/generatePageStub';
import { AppThemeProvider } from 'theme/AppThemeProvider';

export const wrapIntl = (children?: React.ReactNode): JSX.Element => (
  <AppThemeProvider>
    <IntlProvider locale='en'>{children}</IntlProvider>
  </AppThemeProvider>
);
export const wrapDNDIntl = (children?: React.ReactNode): JSX.Element => {
  return <DndProvider backend={HTML5Backend}>{wrapIntl(children)}</DndProvider>;
};

export const wrapPagesProvider = (cardId: string | string[], children: React.ReactNode) => {
  const pages: Record<string, PageMeta> = {};

  const cardIds = Array.isArray(cardId) ? cardId : [cardId];

  cardIds.forEach((_cardId) => {
    pages[_cardId] = pageStubToCreate({ spaceId: v4(), createdBy: v4(), id: _cardId, path: _cardId }) as PageMeta;
  });

  return (
    <PagesContext.Provider
      value={
        {
          pages
        } as any
      }
    >
      {children}
    </PagesContext.Provider>
  );
};

export function mockDOM(): void {
  window.focus = jest.fn();
  document.createRange = () => {
    const range = new Range();
    range.getBoundingClientRect = jest.fn();
    range.getClientRects = () => {
      return {
        item: () => null,
        length: 0,
        [Symbol.iterator]: jest.fn()
      };
    };
    return range;
  };
}
export function mockMatchMedia(result: { matches: boolean }): void {
  // We check if system preference is dark or light theme.
  // This is required to provide it's definition since
  // window.matchMedia doesn't exist in Jest.
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(() => {
      return result;

      // return ({
      //     matches: true,
      // })
    })
  });
}

export function mockStateStore(middleware: Middleware[], state: unknown): MockStoreEnhanced<unknown, unknown> {
  const mockStore = configureStore(middleware);
  return mockStore(state);
}

export type BlocksById<BlockType> = { [key: string]: BlockType };

export function blocksById<BlockType extends Block>(blocks: BlockType[]): BlocksById<BlockType> {
  return blocks.reduce((res, block) => {
    res[block.id] = block;
    return res;
  }, {} as BlocksById<BlockType>);
}
