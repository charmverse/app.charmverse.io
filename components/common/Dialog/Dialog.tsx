import CloseIcon from '@mui/icons-material/Close';
import type { DialogProps as MuiDialogProps } from '@mui/material';
import { Dialog as MuiDialog, DialogContent, DialogTitle, IconButton, DialogActions } from '@mui/material';
import type { ReactNode } from 'react';

import { MobileDialog } from 'components/common/MobileDialog/MobileDialog';
import { useMdScreen } from 'hooks/useMediaScreens';

export type DialogProps = {
  onClose: () => void;
  title: string | ReactNode;
  footerActions?: ReactNode;
  children: ReactNode;
} & Omit<MuiDialogProps, 'title'>;

export function Dialog({ onClose, title, footerActions, children, ...props }: DialogProps) {
  const isMdScreen = useMdScreen();
  const isMobileDialog = !isMdScreen;

  if (isMobileDialog) {
    return (
      <MobileDialog title={title} onClose={onClose} footerActions={footerActions} {...props}>
        {children}
      </MobileDialog>
    );
  }

  return (
    <MuiDialog onClose={onClose} fullWidth maxWidth='sm' {...props}>
      {(!!title || !!onClose) && (
        <DialogTitle sx={{ m: 0, px: 3, pt: 3, pb: 2 }}>
          {title}

          {onClose && (
            <IconButton
              aria-label='close'
              onClick={onClose}
              sx={{
                position: 'absolute',
                right: 8,
                top: 10
              }}
            >
              <CloseIcon color='secondary' />
            </IconButton>
          )}
        </DialogTitle>
      )}

      <DialogContent sx={{ px: 3, display: 'flex', direction: 'column', justifyContent: 'center', flexGrow: 1 }}>
        {children}
      </DialogContent>

      {!!footerActions && <DialogActions sx={{ m: 0, px: 3, py: 2 }}>{footerActions}</DialogActions>}
    </MuiDialog>
  );
}
