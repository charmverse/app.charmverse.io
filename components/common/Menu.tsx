import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import { Menu, MenuItem } from '@mui/material';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import { usePopupState, bindMenu, bindTrigger } from 'material-ui-popup-state/hooks';
import { v4 } from 'uuid';

import Button from 'components/common/Button';
import { StyledListItemText } from 'components/common/StyledListItemText';

export interface MenuOption<T = any> {
  value: T;
  primary: string;
  secondary?: string;
}

interface Props {
  selectedValue: string;
  title?: string;
  loading?: boolean;
  valueUpdated: (value: string) => void;
  options: MenuOption[];
  closeOnSelect?: boolean;
}

export default function SelectMenu ({
  selectedValue,
  valueUpdated,
  options,
  loading = false,
  title,
  closeOnSelect = true } : Props) {

  const popupState = usePopupState({ variant: 'popover', popupId: `menu-${v4()}` });

  const selectedLabel = options.find(opt => opt.value === selectedValue)?.primary;

  return (
    <Box display='block'>
      {
        title && (
          <Typography display='block' variant='caption'>
            {title}
          </Typography>
        )
      }
      <Button
        color='secondary'
        variant='outlined'
        disabled={loading}
        loading={loading}
        endIcon={!loading && <KeyboardArrowDownIcon fontSize='small' />}
        {...bindTrigger(popupState)}
      >
        {selectedLabel}
      </Button>
      <Menu
        {...bindMenu(popupState)}
        PaperProps={{
          sx: { width: 300 }
        }}
      >
        {
            options.map(opt => {

              const { value, primary, secondary } = opt;

              return (
                <MenuItem
                  key={value.primary}
                  selected={selectedValue === value}
                  onClick={() => {
                    valueUpdated(value);
                    if (closeOnSelect) {
                      popupState.close();
                    }
                  }}
                >
                  <StyledListItemText
                    primary={primary}
                    secondary={secondary}
                  />
                </MenuItem>
              );
            })
          }

      </Menu>
    </Box>

  );
}
