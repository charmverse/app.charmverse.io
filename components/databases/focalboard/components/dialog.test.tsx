// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import '@testing-library/jest-dom'
import {render, screen} from '@testing-library/react'

import React from 'react'

import userEvent from '@testing-library/user-event'

import {wrapDNDIntl} from '../testUtils'

import Menu from '../widgets/menu'

import OptionsIcon from '../widgets/icons/options'

import Dialog from './dialog'

describe('components/dialog', () => {
    beforeEach(jest.clearAllMocks)
    test('should match snapshot', () => {
        const {container} = render(wrapDNDIntl(
            <Dialog
                onClose={jest.fn()}
            >
                <div id='test'/>
            </Dialog>,
        ))
        expect(container).toMatchSnapshot()
    })
    test('should return dialog and click onClose button', () => {
        const onCloseMethod = jest.fn()
        render(wrapDNDIntl(
            <Dialog
                onClose={onCloseMethod}
            >
                <div id='test'/>
            </Dialog>,
        ))
        const buttonClose = screen.getByRole('button', {name: 'Close dialog'})
        userEvent.click(buttonClose)
        expect(onCloseMethod).toBeCalledTimes(1)
    })
    test('should return dialog and click to close on wrapper', () => {
        const onCloseMethod = jest.fn()
        const {container} = render(wrapDNDIntl(
            <Dialog
                onClose={onCloseMethod}
            >
                <Menu position='left'>
                    <Menu.Text
                        id='test'
                        icon={<OptionsIcon/>}
                        name='Test'
                        onClick={async () => {
                            jest.fn()
                        }}
                    />
                </Menu>
            </Dialog>,
        ))
        const buttonClose = container.querySelector('.wrapper')!
        userEvent.click(buttonClose)
        expect(onCloseMethod).toBeCalledTimes(1)
    })

    test('should return dialog and click on test button', () => {
        const onTest = jest.fn()
        render(wrapDNDIntl(
            <Dialog
                onClose={jest.fn()}
                toolsMenu={<Menu position='left'>
                    <Menu.Text
                        id='test'
                        icon={<OptionsIcon/>}
                        name='Test'
                        onClick={async () => {
                            onTest()
                        }}
                    />
                </Menu>}
            >
                <div id='test'/>
            </Dialog>,
        ))
        const buttonMenu = screen.getByRole('button', {name: 'menuwrapper'})
        userEvent.click(buttonMenu)
        const buttonTest = screen.getByRole('button', {name: 'Test'})
        userEvent.click(buttonTest)
        expect(onTest).toBeCalledTimes(1)
    })
    test('should return dialog and click on cancel button', () => {
        const {container} = render(wrapDNDIntl(
            <Dialog
                onClose={jest.fn()}
                toolsMenu={<Menu position='left'>
                    <Menu.Text
                        id='test'
                        icon={<OptionsIcon/>}
                        name='Test'
                        onClick={async () => {
                            jest.fn()
                        }}
                    />
                </Menu>}
            >
                <div id='test'/>
            </Dialog>,
        ))
        const buttonMenu = screen.getByRole('button', {name: 'menuwrapper'})
        userEvent.click(buttonMenu)
        const buttonTest = screen.getByRole('button', {name: 'Cancel'})
        userEvent.click(buttonTest)
        expect(container).toMatchSnapshot()
    })
})
