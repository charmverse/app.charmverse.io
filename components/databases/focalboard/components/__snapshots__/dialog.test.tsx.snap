// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/dialog should match snapshot 1`] = `
<div>
  <div
    class="Dialog dialog-back undefined"
  >
    <div
      class="wrapper"
    >
      <div
        class="dialog"
        role="dialog"
      >
        <div
          class="toolbar"
        >
          <button
            aria-label="Close dialog"
            class="Button IconButton IconButton--large"
            title="Close dialog"
            type="button"
          >
            <i
              class="CompassIcon icon-close CloseIcon"
            />
          </button>
        </div>
        <div
          id="test"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`components/dialog should return dialog and click on cancel button 1`] = `
<div>
  <div
    class="Dialog dialog-back undefined"
  >
    <div
      class="wrapper"
    >
      <div
        class="dialog"
        role="dialog"
      >
        <div
          class="toolbar"
        >
          <button
            aria-label="Close dialog"
            class="Button IconButton IconButton--large"
            title="Close dialog"
            type="button"
          >
            <i
              class="CompassIcon icon-close CloseIcon"
            />
          </button>
          <div
            aria-label="menuwrapper"
            class="MenuWrapper"
            role="button"
          >
            <button
              class="Button IconButton IconButton--large"
              type="button"
            >
              <i
                class="CompassIcon icon-dots-horizontal OptionsIcon"
              />
            </button>
          </div>
        </div>
        <div
          id="test"
        />
      </div>
    </div>
  </div>
</div>
`;
